editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,359 @@
+import type {
+  CollabTextFile,
+  CollabTextFileTopLevelElements,
+  CollaborativeEditingSupportSession,
+} from './editor-state'
+import { type CollaborativeEditingSupport } from './editor-state'
+import {
+  writeProjectFileChange,
+  type ProjectChanges,
+  type ProjectFileChange,
+  deletePathChange,
+  ensureDirectoryExistsChange,
+} from './vscode-changes'
+import type { ProjectContentsTree } from '../../../components/assets'
+import {
+  walkContentsTree,
+  type ProjectContentTreeRoot,
+  getProjectFileFromTree,
+  isProjectContentFile,
+  zipContentsTree,
+  addFileToProjectContents,
+} from '../../../components/assets'
+import type { EditorAction, EditorDispatch } from '../action-types'
+import {
+  deleteFile,
+  updateProjectContents,
+  updateTopLevelElements,
+} from '../actions/action-creators'
+import { assertNever, isBrowserEnvironment } from '../../../core/shared/utils'
+import type { ParseSuccess } from '../../../core/shared/project-file-types'
+import {
+  RevisionsState,
+  isTextFile,
+  parseSuccess,
+  textFile,
+  textFileContents,
+} from '../../../core/shared/project-file-types'
+import {
+  ParsedTextFileKeepDeepEquality,
+  TopLevelElementKeepDeepEquality,
+} from './store-deep-equality-instances'
+import * as Y from 'yjs'
+import type { TopLevelElement } from '../../../core/shared/element-template'
+import { isFeatureEnabled } from '../../../utils/feature-switches'
+
+export function collateCollaborativeProjectChanges(
+  oldContents: ProjectContentTreeRoot,
+  newContents: ProjectContentTreeRoot,
+): Array<ProjectFileChange> {
+  let changesToProcess: Array<ProjectFileChange> = []
+
+  function applyChanges(
+    fullPath: string,
+    firstContents: ProjectContentsTree,
+    secondContents: ProjectContentsTree,
+  ): boolean {
+    if (isProjectContentFile(firstContents)) {
+      if (isProjectContentFile(secondContents)) {
+        if (firstContents.content === secondContents.content) {
+          // Do nothing, no change.
+        } else if (isTextFile(firstContents.content) && isTextFile(secondContents.content)) {
+          if (
+            ParsedTextFileKeepDeepEquality(
+              firstContents.content.fileContents.parsed,
+              secondContents.content.fileContents.parsed,
+            ).areEqual
+          ) {
+            // Do nothing, no change.
+          } else {
+            const firstRevisionState = firstContents.content.fileContents.revisionsState
```

Comment:
unused var – is this a problem with the logic that we are not using it? or simply leftover?

Reply:
Just a leftover, fixed in 6d755a897.